New block: Chapter List for handbooks

ryelle opened the pull request:
See #161 — This PR adds the Chapter List block, which is used on the handbook pages to display the navigation in a given handbook. It uses the current post type to list all items in that post type, using the page hierarchy to build up a menu with submenus. 

It also adds styles to the theme to create the three-column layout when the screen is wider than 1440px, and a two col layout from 1200px and up (see https://github.com/WordPress/wporg-developer/issues/161#issuecomment-1406782932).

**Screenshots**

| Small screen (mobile) | Medium (small-mid desktop) | Large (large desktop) |
|---------------|-----------|--------------|
| ![wpcs-small](https://user-images.githubusercontent.com/541093/215607767-ea93d12c-1084-4b18-a41e-25754ec3e079.png) | ![wpcs-mid](https://user-images.githubusercontent.com/541093/215607766-a00287d8-bc1d-4b67-a817-ac2f617edacf.png) | ![wpcs-large](https://user-images.githubusercontent.com/541093/215607762-802c9c62-a9d7-430f-93a1-9ce6341fbb01.png) |

Other handbooks - on production, the Block Editor handbook also uses the page hierarchy, but Plugins & REST API use a navigation menu… so the chapter list on these is a little wonky. For consistency, I'd like to use this same block on all the handbooks, but I'm not sure if there's a reason to use a nav menu instead on these other pages. For now, all are using the Chapter List block.

| Block Editor | Plugins | REST API |
|----------|-----------|--------|
| ![block-editor](https://user-images.githubusercontent.com/541093/215609848-e1dc099a-2ce8-4d03-97f7-1653ab63594e.png) | ![plugins](https://user-images.githubusercontent.com/541093/215609850-8e8377a4-2678-44aa-8148-957e36d37313.png) | ![rest-api](https://user-images.githubusercontent.com/541093/215609855-3e8eb6db-0646-4183-9adc-cbb5957c191e.png) |

**To test**

- Make sure you have handbook content, either by exporting from the production site or by running the parser (`yearn wp-env run cli "wp cron event run --all"` has worked for me).
- View some pages in a handbook
- The menu should appear, and clicking through it should work
- Expanding & collapsing sections should work
- If you land on a nested page, the menu should be open to that page - ex, http://localhost:8888/block-editor/how-to-guides/platform/custom-block-editor/tutorial/

ryelle committed:
Initial chapter list block

ryelle committed:
Try a three-column layout

ryelle committed:
Update style & alignment of ToC and Chapter List

ryelle committed:
Update chapter list block

ryelle commented on source/wp-content/themes/wporg-developer-2023/src/chapter-list/view.js:
```diff
@@ -0,0 +1,44 @@
+			// Create the toggle button.
+			const button = document.createElement( 'button' );
+			button.setAttribute( 'aria-expanded', false );
+			// button.setAttribute( 'aria-label', '' );
```
I meant to follow up with this— this should be something like "open [page name] submenu", will update the PR tomorrow.

StevenDufresne commented on source/wp-content/themes/wporg-developer-2023/theme.json:
```diff
@@ -3,37 +3,79 @@
+			"wporg/table-of-contents": {
+				"color": {
+					"background": "var:preset|color|blueberry-4",
+					"text": "var:preset|color|blueberry-1"
```
Why do this use the `var:` format and the other ones don't? Is there a limitation?

ryelle replied:
No, probably just the way I copied them - swapped in the custom properties 👍🏻

StevenDufresne commented on source/wp-content/themes/wporg-developer-2023/src/style/style.scss:
```diff
@@ -54,6 +51,19 @@ body[class] {
+	.has-three-columns {
+		display: grid;
+		grid-template-columns: 260px 1fr;
+		gap: 20px;
```
Should this use our spacing grid: `var(--wp--preset--spacing--20)`?

StevenDufresne commented:
This is looking sharp!

### Can we reduce the amount of jank on load?
- Left to Right movement
- Top-to-bottom movement.

| Gif |
| --- |
| <img src="https://user-images.githubusercontent.com/1657336/215684620-1b7c2dd2-db76-4bcc-81a6-d91d6db5abf7.gif" width="200" > |

### Nested line alignment

This should be alignment with the middle of the arrow:

<img width="200" alt="Screen Shot 2023-01-31 at 3 27 44 PM" src="https://user-images.githubusercontent.com/1657336/215684890-11fc42dc-36ca-4697-b2c8-8ab7800abfa6.png">

###  Selected item
The designs have a solid line to identify the selected page
| Design | PR |
| --- | --- |
| <img width="292" alt="Screen Shot 2023-01-31 at 3 40 33 PM" src="https://user-images.githubusercontent.com/1657336/215685591-90b6dc20-9849-4251-bd70-f3ed73a264d3.png"> | <img width="200" alt="Screen Shot 2023-01-31 at 3 39 27 PM" src="https://user-images.githubusercontent.com/1657336/215685447-5cdb24e5-58f0-4701-a8b4-68a635d4ec47.png"> |

### Click target
Since the title of the section is a link and doesn't toggle the section, the only way to toggle the section is to click the arrow. The arrow itself is very small and probably requires expert-level motor skills. Can we make it larger?

<img width="200" alt="Screen Shot 2023-01-31 at 3 31 23 PM" src="https://user-images.githubusercontent.com/1657336/215684887-71245c19-04f7-4fb7-bdd7-f6e6cc994644.png">

### Layout

At `1250px` I got this layout:

![localhost_8888_block-editor_reference-guides_richtext_](https://user-images.githubusercontent.com/1657336/215687502-cb91c63f-4cfa-488c-8d36-b72c5b70ea6d.png)

jasmussen commented:
Looks like it can work as a good starting point. The margin looks right on desktop in the last screenshot Steve shared. The margin below the chapters in the first screenshots in this PR, however, looks a little small.

ryelle committed:
Address PR feedback

Use custom properties consistently, fix the style flash when JS loads

ryelle committed:
Increase click target for toggle button

ryelle commented:
> Can we reduce the amount of jank on load?

Good catch, I styled the default list a bit more so it won't shift left/right when the content loads. The collapse shift happens because the menu needs to be expanded to work if the JS doesn't load.

No JS view:
<img width="413" alt="Screenshot 2023-01-31 at 11 33 53 AM" src="https://user-images.githubusercontent.com/541093/215832077-07059048-4668-40e5-90a9-fa7c0bf02f00.png">

> Nested line alignment

Fixed it, though it might appear slightly off on different browsers because it needs to be on a fractional pixel to be centered (even width).

> Selected item

This is a question for design, but the menu as mocked up only has two levels, so the child elements don't need the <kbd>-</kbd> or <kbd>v</kbd> button. Since menus in reality can be arbitrarily deep, all levels use the buttons, and it seemed like having both the button + the border was going to be busy. I'm also not sure how the current border should work on parent items. @WordPress/meta-design What do you think, do we need that border somewhere on here?

<img width="351" alt="Screenshot 2023-01-31 at 12 22 26 PM" src="https://user-images.githubusercontent.com/541093/215835852-8f7f53b6-9ea6-47a1-b9d7-a5f43af97c6f.png">

> Click target

I've made this a bit larger, so it matches the full line-height of the link. It's almost 24px, which is the [recommended minimum target size for WCAG 2.2 AA](https://www.w3.org/WAI/WCAG22/Understanding/target-size-minimum). (WCAG 2.2 is still technically draft, but this is the clearest non-touch target advice I found)

<img width="290" alt="Screenshot 2023-01-31 at 12 23 27 PM" src="https://user-images.githubusercontent.com/541093/215836068-3fd7f61c-62d9-4504-854f-7144203d5a9f.png">

> Layout
> At 1250px I got this layout:

Yeah, from 1200-1440 it's a two column layout. This does need some design iteration, especially on a page like the component or package refs where the menu is _very_ long

<img width="400" alt="" src="https://user-images.githubusercontent.com/541093/215837855-29116036-dca0-4bea-be88-5eab12130ee1.png">

> The margin looks right on desktop in the last screenshot Steve shared. The margin below the chapters in the first screenshots in this PR, however, looks a little small.

The margin there is the `edge-spacing` value, which is what the table of contents box uses — on small screens (at 890px), that size shrinks to 20px. I've updated it to always be 80px, but that looks strange in context with the rest of the margins at 20px. [These are the other spacing values we could use](https://github.com/WordPress/wporg-parent-2021/#spacing), if one of those would be better.

![Screen Shot 2023-01-31 at 12 37 16](https://user-images.githubusercontent.com/541093/215839343-ee08e10b-16e9-45a8-92bb-88b72b39f73f.png)

ryelle committed:
Add the aria label to the toggle buttons

StevenDufresne commented:
Thanks for the updates! 

A couple of minor things that aren't blockers:

### List style with no children
Since we use the blueberry color primarily for actionable elements within content, I wonder if we should consider changing the visual appearance of the "-" in the chapter list. It looks like you would be able to interact with it but it's not an interactive element since it has no children. I've mocked up some examples to illustrate the difference. I don't know. Thinkout out loud. @WordPress/meta-design, @jasmussen 

| Current Design | No dashes | Subdued Dashes | 
| --- | --- | --- |
| <img width="353" alt="Screen Shot 2023-02-01 at 9 16 15 AM" src="https://user-images.githubusercontent.com/1657336/215913670-73e28287-7776-49d7-a9df-88c5b05e2ebc.png"> | <img width="355" alt="Screen Shot 2023-02-01 at 9 16 03 AM" src="https://user-images.githubusercontent.com/1657336/215913675-2c1fb02e-ed39-4339-9b7b-b8e94a9e1a25.png"> |  <img width="358" alt="Screen Shot 2023-02-01 at 9 14 07 AM" src="https://user-images.githubusercontent.com/1657336/215913430-00f82814-22e4-41b6-b46e-2b90fd9d0571.png"> |





### First item

The first item in the chapter list is the name of the handbook and no longer gets a different visual treatment. Should we consider removing it (or removing the dash)? It's currently presented as a sibling of "Get Started" when in fact it's the parent.

| Current | This PR |
| --- | --- |
| <img width="278" alt="Screen Shot 2023-02-01 at 9 02 09 AM" src="https://user-images.githubusercontent.com/1657336/215913070-eeed193b-6574-4c0d-ab6a-8ed82e5ed3c1.png"> | <img width="367" alt="Screen Shot 2023-02-01 at 9 02 04 AM" src="https://user-images.githubusercontent.com/1657336/215913075-ecde060d-e059-4b28-8069-db21fea0ba2d.png"> |

jasmussen commented:
In principle I find myself most attracted to the no dashes option, with less noise. But if you feel strongly, the subdued dashes can work too.

ryelle commented:
> The first item in the chapter list is the name of the handbook and no longer gets a different visual treatment. Should we consider removing it (or removing the dash)? It's currently presented as a sibling of "Get Started" when in fact it's the parent.

That white background/bold text style is the current page indicator, when you're on a different page, there is no visual difference between "Block Editor Handbook" & "Getting Started":

<img width="261" alt="Screenshot 2023-02-01 at 12 11 39 PM" src="https://user-images.githubusercontent.com/541093/216113503-b8b62884-e070-4376-a5c8-75cb257204b1.png">

> In principle I find myself most attracted to the no dashes option, with less noise

Should I make that change, then? It should be just as easy to do either. Personally, I like the subdued dashes better because the spacing looks a little too big with the empty spaces, but I'll leave the final decision to a designer :)

jasmussen commented:
I see your point, and we can go with the subdued dashes. 

It's a bit of a tricky interface, and it might benefit from a full actual tree view, which is that the dashes connect to the vertical line, which is harder for last-child situations on shortening the vertical line.

ryelle committed:
Use a light grey color for the child node dashes

ryelle commented:
Latest block with the lighter dashes:

<img width="274" alt="Screenshot 2023-02-01 at 5 17 59 PM" src="https://user-images.githubusercontent.com/541093/216176609-88b48074-f90b-46a0-b1d5-ff67d712c87c.png">

jasmussen commented:
Sorry for being nitpicky, are those dashes the same gray as the vertical dash? I see them as essentially an ersatz tree view so it'd be good if it was the same gray.

jasmussen commented:
Sorry for being nitpicky, are those dashes the same gray as the vertical dash? I see them as essentially an ersatz tree view so it'd be good if it was the same gray.

ryelle commented:
> Sorry for being nitpicky, are those dashes the same gray as the vertical dash? I see them as essentially an ersatz tree view so it'd be good if it was the same gray.

They technically are, but I think it looks lighter in the screenshot because the dash is 1.5px tall so there's some anti-aliasing going on. I've tweaked the svg a little to avoid the anti-alias, and TBH it still looks slightly lighter to me even though I can color-pick and see they're both #d9d9d9.

<img width="290" alt="Screenshot 2023-02-02 at 11 57 17 AM" src="https://user-images.githubusercontent.com/541093/216391179-d58d6bcb-b4a3-45e6-8e67-21f33652bbe5.png">

I think it's just an optical illusion since the line is shorter? But they are definitely the same now.

jasmussen commented:
We can try this, and revisit. I think the optimization is to do the full tree view look, but it's a less important aspect we can revisit and ticket if need be.

ryelle committed:
Fix sub-pixel rendering of line SVG

ryelle commented:
Sounds good, I'm going to merge this as-is now so we can move forward on this page layout.

I'll leave it up to design to decide if we need to change to a more-connected tree view, and you can open a new issue if so.
